### PR TITLE
Migrating to Typescript

### PR DESCRIPTION
**Pull request by rplankenhorn**
*Description of changes:*
Migrating the codebase to use Typescript instead of vanilla Javascript.  I figured out a number of things about the ask-cli and will add inline comments explaining the different parts of this PR.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: hooks/build.sh [hunk 1]
```diff
@@ -0,0 +1,7 @@
+#!/bin/bash
+
+SCRIPT_DIR="$( cd "$( dirname "$0" )" && pwd )"
+
+cd $SCRIPT_DIR/../lambda/custom
+yarn build
+yarn zip
```
Comment (rplankenhorn) [inline hooks/build.sh:7]: If this script is present, it is executed when running ```ask deploy``` and it expects a ```build.zip``` artifact to be created.  If this script doesn't create it, ```ask deploy``` will fail.

File: lambda/custom/package.json [hunk 2]
```diff
@@ -5,7 +5,17 @@
   "main": "index.js",
   "scripts": {
     "setup": "bash ../../bin/setup.sh",
-    "test": "echo \"Error: no test specified\" && exit 1"
+    "test": "echo \"Error: no test specified\" && exit 1",
+    "clean": "rimraf .build",
+    "compile": "tsc --build tsconfig.json --pretty",
+    "build-final": "cpy package.json .build && cpy yarn.lock .build && cpy .env .build && cd .build/ && yarn install --production=true",
```
Comment (rplankenhorn) [inline lambda/custom/package.json:11]: Now that we can control which files are copied into the bundle, I was able to re-enable ```dotenv```.

File: .travis.yml [hunk 3]
```diff
@@ -11,6 +11,7 @@ before_install:
 install:
   - cd lambda/custom
   - yarn
+  - cd ../..
```
Comment (cdubz) [inline .travis.yml:14]: I was actually going to suggest adding a `package.json` to the repo root as well and having it alias the commands in `lamdba/custom` so the `yarn` stuff can be run from one place (and the same place as `ask deploy`). But that can come in another PR (:

Comment (rplankenhorn) [reply .travis.yml:14]: Great idea!  I'll do some experimenting with it.

**Pull request merged**